### PR TITLE
core: introduce (and fix) code to pass errno to other OSes

### PR DESCRIPTION
**Pull request by wjwithagen**
- Some of the errno conversions were lost in the conversion
   to Cmake config
 - rename ceph_to_host_errno to _ceph_to_hostos_errno
   to indicate that the errnos are define per OS.
 
This is the second part, creating the basics.
Next it needs to be glued into communication with the peers.

Signed-off-by: Willem Jan Withagen <wjw@digiware.nl>

### CODE REVIEW

File: src/include/types.h [hunk 1]
```diff
@@ -493,11 +495,12 @@ struct errorcode32_t {
   }
 
   void encode(bufferlist &bl) const {
-    ::encode(code, bl);
+    __s32 newcode = hostos_to_ceph_errno(code); 
+    ::encode(newcode, bl);
```
Comment (liewegas) [inline src/include/types.h]: This was teh weird missing piece in the earlier version. Do you mind separating it into a separate patch?  Eveything else is fixing the build and cmake-related regresssions.  This is a change in behavior so that the the server-side can also use non-Linux errno values.

Comment (wjwithagen) [inline src/include/types.h]: @liewegas 
Sure, no problem.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -538,6 +538,24 @@ set(libcommon_files
   ${auth_files}
   ${mds_files})
 
+if(FREEBSD)
+  list(APPEND libcommon_files common/freebsd_errno.cc)
+endif()
```
Comment (tchaikov) [inline src/CMakeLists.txt]: nit, could use `elseif` as they are mutual exclusive.

Comment (wjwithagen) [inline src/CMakeLists.txt]: @tchaikov 
Good suggestion, will take it along.

File: src/common/freebsd_errno.cc [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+  return r; // otherwise return original value
+}
+
+// Build a table with the FreeBSD error as index
+// and the Linux error as value
+// Use the fact that the arry is initialised per default on all 0's
+// And we do not translate for 0's, but return the original value.
+static __s32 hostos_to_ceph_conv[256] = {
+	//        FreeBSD errno Linux errno
+	H2C_ERRNO(EDEADLK,	35),   	/* Resource deadlock avoided */
```
Comment (tchaikov) [inline src/common/freebsd_errno.cc]: i think
```c++
[EDEADL] = 35,
```
is more straightforward.

File: src/common/freebsd_errno.cc [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+// converts Host OS errno values to linux/Ceph values
+__s32 hostos_to_ceph_errno(__s32 r)
+{
+  int sign = (r<0?-1:1);
+  int err = abs(r);
+  if (err <256 && hostos_to_ceph_conv[err] !=0 ) {
+    err = hostos_to_ceph_conv[err];
+  }
+  return err * sign;
```
Comment (tchaikov) [inline src/common/freebsd_errno.cc]: please note that if `errno <= 34`, we just use its original value.

```c++
__s32 hostos_to_ceph_errno(__s32 r)
{
  int err = std::abs(r);
  if (err > 34 && err < 256) {
    err = hostos_to_ceph_conv[err];
  }
  return std::copysign(err, r);
}

File: src/common/freebsd_errno.cc [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+    }
+  }
+  return r; // otherwise return original value
+}
+
+// Build a table with the FreeBSD error as index
+// and the Linux error as value
+// Use the fact that the arry is initialised per default on all 0's
+// And we do not translate for 0's, but return the original value.
+static __s32 hostos_to_ceph_conv[256] = {
```
Comment (wjwithagen) [inline src/common/freebsd_errno.cc]: @tchaikov 
Will do

File: src/common/freebsd_errno.cc (hunk 3 above)
Comment (wjwithagen) [inline src/common/freebsd_errno.cc]: @tchaikov 
I like the macro more, since modifying the operational code is than just replacing the macro.
BTW We need to use the FreeBSD errno defines, and that does not have `EDEADL`

File: src/common/freebsd_errno.cc (hunk 4 above)
Comment (wjwithagen) [inline src/common/freebsd_errno.cc]: @tchaikov 
Not quite.... 
```
#define EDEADLK         11              /* Resource deadlock avoided */
                                        /* 11 was EAGAIN */
```
So those got swapped in FreeBSD.
And all values that are not modified are 0, and are thus passed without modification.
So no special need to check for `< 34`
But we do need to check if the replacement is `!= 0` before replacing.

Now if the array conversion is to your liking, I would also change the original function: `ceph_to_hostos_errno()` to work in the same way.

**Comment by wjwithagen** [comment]
@liewegas @tchaikov 
This should be more or less the PR for this, plus/minus some rebasing and or change requests.

File: src/common/freebsd_errno.cc [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        H2C_ERRNO(ENOTRECOVERABLE, 131),/* State not recoverable */
+        H2C_ERRNO(EOWNERDEAD,	130),	/* Previous owner died */
+	};
+
+// converts from linux errno values to host values
+__s32 ceph_to_hostos_errno(__s32 r)
+{
+  int sign = (r<0?-1:1);
+  int err = std::abs(r);
+  if (err <256 && hostos_to_ceph_conv[err] !=0 ) {
```
Comment (tchaikov) [inline src/common/freebsd_errno.cc]: add space around operator.

File: src/common/freebsd_errno.cc [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        H2C_ERRNO(ENOTCAPABLE,	EPERM),	/* Capabilities insufficient */
+        H2C_ERRNO(ECAPMODE,	EPERM),	/* Not permitted in capability mode */
+        H2C_ERRNO(ENOTRECOVERABLE, 131),/* State not recoverable */
+        H2C_ERRNO(EOWNERDEAD,	130),	/* Previous owner died */
+	};
+
+// converts from linux errno values to host values
+__s32 ceph_to_hostos_errno(__s32 r)
+{
+  int sign = (r<0?-1:1);
```
Comment (tchaikov) [inline src/common/freebsd_errno.cc]: no need to calc the sign. `std::copysign(err, r);` would suffice

File: src/common/freebsd_errno.cc [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+  if (err <256 && hostos_to_ceph_conv[err] !=0 ) {
+    err = ceph_to_hostos_conv[err];
+  }
+  return err * sign;
+}
+
+// converts Host OS errno values to linux/Ceph values
+__s32 hostos_to_ceph_errno(__s32 r)
+{
+  int sign = (r<0?-1:1);
```
Comment (tchaikov) [inline src/common/freebsd_errno.cc]: ditto.

File: src/common/freebsd_errno.cc (hunk 7 above)
Comment (wjwithagen) [inline src/common/freebsd_errno.cc]: @tchaikov 
I looked at that function, but it seemed to me that it was floating point only
And mostly delt with things as `NAN`.
OTOH, the implementation does exactly what I do.

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-16_18:45:23-rados-wip-sage-testing-distro-basic-smithi/

**Comment by wjwithagen** [comment]
@liewegas 
Arrgh, still a syntax error sneaked in... :(
Will submit fix
